### PR TITLE
Tech debt: refactor money, percentage, and decimal widgets

### PR DESCRIPTION
**Pull request by BCerki**
Card: https://app.zenhub.com/workspaces/climate-action-secretariat-60ca4121764d710011481ca2/issues/bcgov/cas-cif/959

### CODE REVIEW

File: app/lib/theme/widgets/DecimalWidget.tsx [hunk 1]
```diff
@@ -1,24 +1,34 @@
 import { WidgetProps } from "@rjsf/core";
 import NumberFormat from "react-number-format";
 
-export const MoneyWidget: React.FC<WidgetProps> = ({
+export const DecimalWidget: React.FC<WidgetProps> = ({
```
Comment (BCerki) [inline app/lib/theme/widgets/DecimalWidget.tsx:4]: I'm not confident `DecimalWidget` is the best name for this--`NumberWidget`?

File: app/lib/theme/widgets/DecimalWidget.tsx [hunk 2]
```diff
@@ -1,24 +1,34 @@
...
   onChange,
-  value,
 }) => {
+  // If we are using this widget to show numbers as money or percent, we can set `isMoney` or `isPercentage` to true in the uiSchema.
+  const isMoney = uiSchema?.isMoney;
+  const isPercentage = uiSchema?.isPercentage;
+
+  // If we need to set the amount of decimal places, we can set it in the uiSchema, otherwise it will default to 10
+  const numberOfDecimalPlaces =
+    isMoney || isPercentage ? 2 : uiSchema.numberOfDecimalPlaces ?? 10;
```
Comment (BCerki) [inline app/lib/theme/widgets/DecimalWidget.tsx:19]: 10 is arbitrary

File: app/lib/theme/widgets/DecimalWidget.tsx [hunk 3]
```diff
@@ -29,7 +39,7 @@ export const MoneyWidget: React.FC<WidgetProps> = ({
           ) {
```
Comment (BCerki) [inline app/lib/theme/widgets/DecimalWidget.tsx:39]: `PercentageWidget` additionally had a check to make sure the number was between 0-100, but I think there could be cases (e.g. emissions calculations) where a user might need to enter a value outside of 0-100, so I took it out.

File: app/lib/theme/widgets/DecimalWidget.tsx [hunk 4]
```diff
@@ -10,17 +10,38 @@ export const DecimalWidget: React.FC<WidgetProps> = ({
   label,
```
Comment (BCerki) [inline app/lib/theme/widgets/DecimalWidget.tsx:10]: I'm not confident `DecimalWidget` is the best name for this widget--`NumberWidget`?

File: app/lib/theme/widgets/NumberWidget.tsx [hunk 5]
```diff
@@ -29,7 +42,7 @@ export const MoneyWidget: React.FC<WidgetProps> = ({
           ) {
```
Comment (BCerki) [inline app/lib/theme/widgets/NumberWidget.tsx:42]: PercentageWidget additionally had a check to make sure the number was between 0-100, but I think there could be cases (e.g. emissions calculations) where the value might be outside of 0-100, so I took it out.

Comment (pbastia) [reply app/lib/theme/widgets/NumberWidget.tsx:42]: Very nice! Percentage values aren't necessarily bound by 0-100, and that concern should be up to the user of the widget instead.

File: app/lib/theme/widgets/NumberWidget.tsx [hunk 6]
```diff
@@ -1,24 +1,37 @@
...
+  // If we are using this widget to show numbers as money or percent, we can set `isMoney` or `isPercentage` to true in the uiSchema.
+  const isMoney = uiSchema?.isMoney;
+  const isPercentage = uiSchema?.isPercentage;
+
+  // If we need to set the amount of decimal places, we can set it in the uiSchema, otherwise there will be no decimal places.
+  const numberOfDecimalPlaces = uiSchema.numberOfDecimalPlaces
+    ? uiSchema.numberOfDecimalPlaces
+    : isMoney
+    ? 2
+    : 0;
```
Comment (pbastia) [inline app/lib/theme/widgets/NumberWidget.tsx:22]: In the diff component, we enforce money to have 2 decimal places, whereas here we allow the user to configure it. I'd be in favour of defaulting to 2 decimal places for money no matter what

File: app/lib/theme/widgets/ReadOnlyAdjustableCalculatedValueWidget.tsx [hunk 7]
```diff
@@ -22,11 +22,11 @@ const ReadOnlyAdjustableCalculatedValueWidget: React.FC<WidgetProps> = (
...
             <NumberFormat
               thousandSeparator
-              fixedDecimalScale={isMoney || isPercent}
+              fixedDecimalScale={isMoney || isPercentage}
               id={id}
               prefix={isMoney ? "$" : ""}
-              suffix={isPercent ? "%" : ""}
-              decimalScale={isMoney || isPercent ? 2 : 10}
+              suffix={isPercentage ? "%" : ""}
+              decimalScale={isMoney || isPercentage ? 2 : 10}
```
Comment (pbastia) [inline app/lib/theme/widgets/ReadOnlyAdjustableCalculatedValueWidget.tsx:29]: We can probably add support for configurable decimal places from the uiSchema in a further refactor

File: app/lib/theme/widgets/ReadOnlyNumberWidget.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  // If we are using this widget to show numbers as money or percent, we can set `isMoney` or `isPercentage` to true in the uiSchema.
+  const isMoney = uiSchema?.isMoney;
+  const isPercentage = uiSchema?.isPercentage;
+
+  // If we need to set the amount of decimal places, we can set it in the uiSchema, otherwise there will be no decimal places.
+  const numberOfDecimalPlaces = uiSchema.numberOfDecimalPlaces
+    ? uiSchema.numberOfDecimalPlaces
+    : isMoney
+    ? 2
+    : 0;
```
Comment (pbastia) [inline app/lib/theme/widgets/ReadOnlyNumberWidget.tsx:18]: Same comment about decimal places for money

**Review by pbastia — CHANGES_REQUESTED** [review]
Very nice! a small change about decimal places for money items

File: app/lib/theme/widgets/ReadOnlyAdjustableCalculatedValueWidget.tsx (hunk 7 above)
Comment (BCerki) [reply app/lib/theme/widgets/ReadOnlyAdjustableCalculatedValueWidget.tsx:29]: I mostly already added it; I'll add it quickly here

File: app/lib/theme/CustomDiffFields.tsx [hunk 9]
```diff
@@ -98,12 +98,15 @@ const showNumberDiff = (
...
-  isPercentage: boolean
+  isPercentage: boolean,
+  numberOfDecimalPlaces: number = 0
 ) => (
   <>
     <span id={id && `${id}-diffOld`} className="diffOld">
       <NumberFormat
         thousandSeparator
+        fixedDecimalScale={true}
+        decimalScale={isMoney ? 2 : numberOfDecimalPlaces ?? 0}
```
Comment (matthieu-foucault) [inline app/lib/theme/CustomDiffFields.tsx:109]: could use some parentheses, also this logic should be written once and stored in a variable

Comment (BCerki) [reply app/lib/theme/CustomDiffFields.tsx]: My formatter took out the parentheses

**Review by Sepehr-Sobhani — APPROVED** [review]

**Pull request merged**